File: Algorithms/1338_Reduce_Array_Size_to_The_Half/Java/ReduceArraySizetoTheHalfSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int minSetSize(int[] arr) {
        Map<Integer, Integer> counter = new HashMap<>();
        List<Integer> tempList = new ArrayList<>();
        int res = arr.length, resCount = 0;

        for (int i : arr) {
            if (counter.containsKey(i)) {
                counter.put(i, counter.get(i) + 1);
            } else {
                counter.put(i, 1);
                tempList.add(i);
            }
        }
        tempList.sort((x1, x2) -> counter.get(x2) - counter.get(x1));

        for (int i : tempList) {
            resCount += 1;
            res -= counter.get(i);
            if (res <= arr.length / 2) return resCount;
        }
        return resCount;
    }
}

File: Algorithms/1338_Reduce_Array_Size_to_The_Half/Python/Reduce_Array_Size_to_The_Half_Solution_1.py

# Space: O(n)
# Time: O(n)

import collections


class Solution:
    def minSetSize(self, arr):
        length = len(arr)
        counter = collections.Counter(arr)
        cache = sorted(list(set(arr)), key=lambda x: -counter[x])
        res, res_counter = length, 0

        for i in cache:
            res -= counter[i]
            res_counter += 1
            if res <= length // 2: return res_counter
        return res_counter




